utils/s2n_random.c

```diff
@@ -332,8 +341,84 @@ S2N_RESULT s2n_get_private_random_bytes_used(uint64_t *bytes_used)
     return S2N_RESULT_OK;
 }
 
-static int s2n_rand_urandom_impl(void *ptr, uint32_t size)
+S2N_RESULT s2n_rand_get_dev_urandom(struct s2n_rand_device **device)
+{
+    RESULT_ENSURE_REF(device);
+    RESULT_ENSURE(s2n_in_unit_test(), S2N_ERR_NOT_IN_UNIT_TEST);
+    *device = &s2n_dev_urandom;
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_rand_device_open(struct s2n_rand_device *device)
+{
+    RESULT_ENSURE_REF(device);
+    RESULT_ENSURE_REF(device->source);
+
+    DEFER_CLEANUP(int fd = -1, s2n_rand_entropy_fd_close_ptr);
+    while (fd < 0) {
+        fd = open(device->source, ENTROPY_FLAGS);
+        if (fd < 0 && errno != EINTR) {
+            RESULT_BAIL(S2N_ERR_OPEN_RANDOM);
+        }
+    }
+
+    struct stat st = { 0 };
+    RESULT_ENSURE(fstat(fd, &st) == 0, S2N_ERR_OPEN_RANDOM);
+    device->dev = st.st_dev;
+    device->ino = st.st_ino;
+    device->mode = st.st_mode;
+    device->rdev = st.st_rdev;
+
+    device->fd = fd;
+
+    /* Disable closing the file descriptor with defer cleanup */
+    fd = UNINITIALIZED_ENTROPY_FD;
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_rand_device_validate(struct s2n_rand_device *device)
 {
+    RESULT_ENSURE_REF(device);
+    RESULT_ENSURE_NE(device->fd, UNINITIALIZED_ENTROPY_FD);
+
+    /* Ensure that the random device is still valid by comparing it to the current file descriptor
+     * status. From:
+     * https://github.com/openssl/openssl/blob/260d97229c467d17934ca3e2e0455b1b5c0994a6/providers/implementations/rands/seeding/rand_unix.c#L513
+     */
+    struct stat st = { 0 };
+    RESULT_ENSURE(fstat(device->fd, &st) == 0, S2N_ERR_OPEN_RANDOM);
+    RESULT_ENSURE_EQ(device->dev, st.st_dev);
+    RESULT_ENSURE_EQ(device->ino, st.st_ino);
+    RESULT_ENSURE_EQ(device->rdev, st.st_rdev);
+
+    /* Ensure that the mode is the same (equal to 0 when xor'd), but don't check the permission bits. */
+    mode_t permission_mask = ~(S_IRWXU | S_IRWXG | S_IRWXO);
+    RESULT_ENSURE_EQ((device->mode ^ st.st_mode) & permission_mask, 0);
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_rand_device_get_valid_fd(struct s2n_rand_device *device, int *fd)
+{
+    RESULT_ENSURE_REF(device);
+    RESULT_ENSURE_REF(fd);
+
+    if (s2n_result_is_error(s2n_rand_device_validate(device))) {
+        RESULT_GUARD(s2n_rand_device_open(device));
+    }
```

Comment:
How does this behave with chroot?

Reply:
I tested this locally by calling s2n_init, invalidating the file descriptor, calling chroot, invalidating the file descriptor, and then calling `s2n_get_public_random_data()`.`s2n_rand_device_validate()` failed since the file descriptor was changed, and then /dev/urandom was attempted to be opened. Since /dev/urandom didn't exist in the chroot tree, opening /dev/urandom failed with S2N_ERR_OPEN_RANDOM:
```
Error Message: 'error opening urandom'
 Debug String: 'Error encountered in /Users/vclarksa/w/s2n-tls-fork/./utils/s2n_random.c:352'
 System Error: No such file or directory (2)
```

Calling chroot requires elevated permissions, which makes it difficult to add to a normal unit test. After our testing infrastructure supports using chroot, a test for this can be added (https://github.com/aws/s2n-tls/issues/4360).